### ISSUE TITLE
Add extension types to default .editorconfig templates

### ISSUE THREAD

**Issue by timheuer** [dotnet/roslyn#47787]
In the `.editorconfig` templates one can add through Visual Studio, the file types registered for the tab spacing are:
```
indent_style = tab
# Code files
[*.{cs,csx,vb,vbx}]
indent_size = 4
insert_final_newline = true
charset = utf-8-bom
```

Presently in Visual Studio, the XML editor *defaults* are tabs, not spaces like all other areas.  EditorConfig is some guidance we use for consistency and we should add common .NET XML file types to our default template.  Namely change to add csproj/vbproj/fsproj/xml as types so the new version would be something like this.

```
indent_style = tab
# Code files
[*.{cs,csx,vb,vbx,csproj,vbproj,fsproj,xml}]
indent_size = 4
insert_final_newline = true
charset = utf-8-bom
```

**Comment by sharwell** [comment]
@timheuer There are several different ways to create a **.editorconfig** file, and they are owned by several different teams. Can you include a screenshot showing how you are creating the file so we can get this issue assigned correctly?

**Comment by davkean** [comment]
And .props/.targets.

**Comment by jmarolf** [comment]
> Presently in Visual Studio, the XML editor defaults are tabs, not spaces like all other areas. EditorConfig is some guidance we use for consistency and we should add common .NET XML file types to our default template. Namely change to add csproj/vbproj/fsproj/xml as types so the new version would be something like this.

Personally, agree with this. Would like us to at least have the editorconfig files we generate support spaces as that is the default for .NET code files today.

> There are several different ways to create a .editorconfig file, and they are owned by several different teams. Can you include a screenshot showing how you are creating the file so we can get this issue assigned correctly?

All the different teams are creating editorcondig files in accordance with how we've advised them so I don't think we need more info to make a call here on whether this is the right thing to do.

> they are owned by several different teams.

- Visual Studio "Add New Item" template is defined [here](https://github.com/dotnet/templates/blob/4f0aa063ac699e25144895c469b542835eb93282/src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/TemplateConstants.cs#L12).
- Tools->Options "Generate .editorconfig file from settings" is defined [here](https://github.com/dotnet/roslyn/blob/7e89bce35497ac106ed58621a281598bd2d6b2af/src/Workspaces/Core/Portable/Options/EditorConfig/EditorConfigFileGenerator.cs#L18)
- Intellicode editorconfig generator (internal) is [here](https://devdiv.visualstudio.com/DevDiv/_git/Intellicode-EditorConfig?path=%2FEditorConfigGenerator.Core%2FStyles%2FStyleAggregator.cs&version=GBmaster&line=22&lineEnd=23&lineStartColumn=1&lineEndColumn=1&lineStyle=plain)

Would not be a lot of work to update all of them to support this if we wanted. I still think the .NET Tools group (us) should make that call on what the defaults for .NET should be. These other editorconfig impls should follow whatever desgin we set out.

**Comment by CyrusNajmabadi** [comment]
I think we can only make decisions on the xml experience by discussing this with the xml team.  Not sure how is best tehre.  But maybe @mgoertz-msft has an idea?

**Comment by mgoertz-msft** [comment]
Plain XML is not actually owned by our team. Interestingly, if I reset my VS settings to either General or C# I end up with TabSize of 4 and using tabs but what I had before used to be 2 spaces. Not sure where that came from but the .editorconfig template [here ](https://github.com/dotnet/templates/blob/4f0aa063ac699e25144895c469b542835eb93282/src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/TemplateConstants.cs#L12)also has a TabSize of 2.

Anyway, I would have to find the owning team but I'm not sure that they will have a strong opinion since the XML editor is pretty much in maintenance mode.

And for XAML, it would make sense to add it to that list too since 4 spaces matches the default anyway.

**Comment by timheuer** [comment]
@mgoertz-msft - what I'm asking here is to change the templates that are the .NET team.  The defaults from VS are being looked at by the XML editor team.

**Comment by genlu** [comment]
Design meeting note:

- Add csproj/vbproj, but don't include fsproj and xml.
- Generated editorconfig will contains new file types (tools-> options)

**Comment by mgoertz-msft** [comment]
@timheuer Yeah, I get that. But since .editorconfig settings take precedence over VS defaults it could still be confusing to some users if their defaults for other languages change once they add a new .editorconfig to a folder.

**Comment by timheuer** [comment]
@mgoertz-msft but that's no diff than the overriding we would do for the other ones (e.g., I could have changed my C# to tabs and this editorconfig changes my defaults).  The goal here is that once a .NET template is added that the most used types for that project follow the same.

**Comment by mgoertz-msft** [comment]
@timheuer Sure, but I would argue that for the best user experience the default template (not the dynamically generated one obviously) should be in sync with the defaults and not competing with them. It's a discussion that should be had with the XML team.

For XAML on the other hand it's all groovy. :-)

**Comment by jmarolf** [comment]
> But since .editorconfig settings take precedence over VS defaults it could still be confusing to some users if their defaults for other languages change once they add a new .editorconfig to a folde

This is why we are only going to change `csproj` and `vbproj` files by default. all other filetypes could have multiple purposes but we know these files are only about .NET tooling

The proposal is to change what we generate for editorconfig files from this:
```ini
root = true
# All files
[*]
indent_style = space
# Code files
[*.{cs,csx,vb,vbx}]
indent_size = 4
insert_final_newline = true
charset = utf-8-bom
```

to this:
```ini
root = true
# All files
[*]
indent_style = space
# Project Files
[*.{csproj,vbproj}]
indent_size = 2
# Code files
[*.{cs,csx,vb,vbx}]
indent_size = 4
# Project and Code Files
[*.{cs,csx,csproj,vb,vbx,vbproj}]
insert_final_newline = true
charset = utf-8-bom
```

This matches the spacing that is used today in csproj files when they are created via the .NET CLI or Visual Studio.

@timheuer any objections?

**Comment by timheuer** [comment]
@jmarolf no objections on the isolated change.  I'm already working with XML editor on their VS defaults.

**Comment by davkean** [comment]
It should respect options, no?

**Comment by jmarolf** [comment]
If the XML editor doesn't respect the setting then thats definitely a bug.

**Comment by davkean** [comment]
No, I mean I've told XML files to have 4 spaces in options, and you're explicitly setting it to 2.

**Comment by jmarolf** [comment]
ah, well my intuition is that for the template case the user is saying "Override my defaults with whatever is recommended" in that case I am ok with setting two spaces even the user has different VS settings.

For the other two cases 1.) intellicode 2.) generating and editorconfig file from existing settings you are correct. The User just wants an editorconfig file that best represents their current settings.  In this case we should just output whatever the VS Settings are. 

perhaps the intellicode inference could do more advanced things but this seems a reasonable place to start.
